doc/developer/design/20210401_volatile_sources.md

```diff
@@ -0,0 +1,294 @@
+# Volatile sources
+
+## Summary
+
+Our users want to connect Materialize to upstream sources that do not provide
+guarantees about what data they will present. These sources are in tension with
+the goal of providing a deterministic, replayable system.
+
+This design doc proposes to:
+
+  * Introduce "volatile source" as a term of art that describes a source that
+    does not have upstream durability.
+
+  * Use the terminology consistently in the docs site and error messages to
+    educate our users about volatile sources.
+
+  * Restrict the use of volatile sources in contexts where they could impede
+    future improvements to the system's determinism and correctness.
+
+## Goals
+
+The only goal is to **stabilize the [PubNub] and [SSE] sources**.
+
+Both of these are volatile sources with some worrying correctness properties,
+but their existence will massively improve the "wow factor" for demos of
+Materialize Cloud.
+
+[PubNub]: https://github.com/MaterializeInc/materialize/pull/6189
+[SSE]: https://github.com/MaterializeInc/materialize/pull/6208
+
+## Description
+
+### Terminology
+
+A "volatile source" is any source that cannot reliabily present the same data
+with the same timestamps across different instantiations.
```

Comment:
I like this definition but I think you might want to tease out that "instantiation" includes

- each materialized view/index using that source
- each ad-hoc query using that source
- each tail using that source
- each time Materialize restarts, everything is reinstantiated

because the different cases where Materialize creates a new "instance" of a source may not be obvious to folks

Reply:
Took a stab at this.